Upload csv 54

ice1080 opened the pull request:
Hey @ddfridley this issue/component(s) should be fully finished. Please let me know if you need me to add or change anything for this!

ddfridley commented on stories/upload-csv-popup.stories.js:
```diff
@@ -0,0 +1,219 @@
+ExistingTableMatchEmail.args = { ...existingTableNewEmailArgs }
+ExistingTableMatchEmail.play = async ({ canvasElement }) => {
+    const canvas = within(canvasElement)
+    await new Promise(r => setTimeout(r, 1000))
```
In these tests, am I suppose to have a csv file or something to upload.  A demo of this might help, or we may have to find a way to add csv files for testing somewhere in the repo.  Not sure if there are differences between your environment and mine that make a difference.

ice1080 replied:
There's an issue with these where the test uploads a file and the story continues in the background, but the browser opens up the "browse for file" dialogue. The story itself still works from what I saw on my end, and I spent some time trying to get the dialogue to go away but didn't want to spend a ton of time on it since the functionality was there and it was an issue specific to storybook. I'll do another pass through on it to see if I can find a better solution.

ice1080 replied:
Actually, I'm rerunning some of these they are no longer updating the electionObj so something's definitely not working quite right. I swear this was just working... I'll get that fixed up too.

ice1080 replied:
I do see these stories working, but the dialog popping up appears to be a defect or issue within storybook from what I can tell. I will create an issue on their github page in the meantime, and add a decorator to the stories to explain this behavior to future people using these stories.

ddfridley reviewed (changes requested):
This is really awesome and I see you have done a lot of work, and I see you read the react component guidelines and I appreciate that.  The change request are pretty minor compare to all the work you have done.  It's really appreciated.

ice1080 commented on bashrcsetup.sh:
```diff
@@ -8,4 +8,5 @@ echo "alias branches=\"git for-each-ref --sort=committerdate refs/heads/ --forma
 echo "alias dbup=\"export MONGODB_URI=mongodb://localhost:27017/test?connectTimeoutMS=3000000 && mkdir -p tmp/db && rm -rf tmp/db/* && npm run db-up\"" >> .bashrc
-echo "alias dbdown=\"npm run db-down\"" >> .bashrc
\ No newline at end of file
+echo "alias dbuplinux=\"export MONGODB_URI=mongodb://localhost:27017/test?connectTimeoutMS=3000000 && sudo rm -rf tmp/db && mkdir -p tmp/db && npm run db-up-linux\"" >> .bashrc
```
I'll work on this. I enjoy bash scripting and am sure that I can get something working for multiple os's.

ice1080 commented on app/components/upload-csv.js:
```diff
@@ -0,0 +1,89 @@
+    popup: {
+        position: 'absolute',
+        top: 'calc(100% + 0.75rem)',
+        transform: 'translateX(-25%)',
```
I'll try to find a more elegant solution to the hardcode. I believe I have to get the width of either the sub component or the super component, can't remember which, and not sure how easy that will be, but I'll definitely look into it. In terms of the positioning, I was kindof basing this on figma and that this button would only be displayed when there's a side bar as well, so we wouldn't have to worry about the popup getting cutoff, but that's probably not the best assumption. I'll work on something better. And I would assume the messed up padding is from the same issue; I believe that flexbox is trying to shift the file-drop to the right as much as it can, but I'll see if that gets fixed when I fix the cutoff issue.
I'll also add a story to the candidate table similar to the upload csv story to fully test an upload and electionObj update with all of it rendered. 
Thanks @ddfridley for the review, let me know if there's anything else!

ice1080 commented:
Just pushed a fix for these items. 
~
For the bash setup stuff, I removed the npm commands and moved everything to the bashrc setup file. Let me know if you'd rather go the route of a script to run the db and I can do that instead. 
~
For the location of the popup, it seemed like you were wanting the popup centered on the entire screen, so I did that with both the width and height. I added these popup variables to the theme in order to use them in both places they were needed, but also felt like that might not be the best use of the theme file. Let me know if you want me to switch those back to hardcoded values.
~
And lastly for the stories, I fixed the display issue in the candidate table and added a note to the stories explaining the storybook issue. 
~
Let me know if there's anything else I can do for this, or if you want me to look into other options for any of the above! Thanks @ddfridley

ddfridley commented on app/components/upload-csv-popup.js:
```diff
@@ -0,0 +1,431 @@
+    fileIcon: {
+        width: '4rem',
+        height: '4rem',
+        margin: '20px 0',
```
Instead of 20px can you use 1.25rem

ice1080 replied:
Absolutely. I'll get the margin fixed up and look into your file. It's definitely helpful having a more full csv like this, I hadn't seen any others in the project yet. I did put various errors in but will do my best to make sure that at least one error always shows up if the election obj doesn't get updated, and also will handle your case. Thanks!

Also sorry I wasn't able to join this past Tuesday! I was really hoping to but had something else that night.

ice1080 replied:
Looks like I had that working previously but then when I added some other stuff it must have broken. 
As part of the next component I may try to get jest working with storybook testing library so that there are a form of unit tests around all of this. It should help speed up development and ensure nothing breaks in the future too.
Just pushed a fix for all of the above! I did some more testing around it all and it seems to be good now. I added your csv file as a story titled "Other Data File" just so that it's recorded as well. Let me know if you see anything else I can update! 
(also just merged upstream main into this branch)

ddfridley reviewed (changes requested):
This is awesome.  I went through all the test cases and I see that if I close the file selector pop-up that I see the result.

I have one issue though, when I go to the Candidate Table the Default story, and I try to upload this file:

upload-test.csv
```
Ballot Name,Office,Slate,Email Address,Type,unique_id,recorder_url,recorder_url_updated,viewer_url,viewer_url_updated
Breeze Velazquez,Academic Affairs Commission,For the People,breezey06@fake.enciv.org,candidate,5e99f99a4cc6803ba8c1864f,,,,
Alice Naland,Campus Events Commission,Independent,alicenaland@fake.enciv.org,candidate,5e99f99a4cc6803ba8c18651,,,,
```

After I select the file and hit [Extract Data] I don't see anything.  I assume that my column names don't match, but there is no indication of that.  I think in the Upload CSV Popup / File Missing Headers I don't see any error messages either.

- [ ] There should be an error message, even better if there was a list of headers that weren't found but required.

- [ ] Also, I tried the check sample link, and I know it won't work without the server working, but I wish it would open in a separate tab.

I agree with moving dbup and down entirely into .bashrc  Originally I wanted it to be an npm script so it could run in a continuous integration environment, but now I've learned how to start an in memory version of Mongo from within Jest and that's wonderful.  

Also, I didn't know you could put decorators at the top of .storeies files, thanks for shown me.


Thanks!

ddfridley reviewed (approved):
Looks really great!
